### PR TITLE
Fix crash in FAR helpers when only part of a linked-document set has been updated

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.cs [hunk 1]
```diff
@@ -271,7 +272,15 @@ public static IEnumerable<TSymbol> FindSimilarSymbols<TSymbol>(TSymbol symbol, C
                 foreach (var linkedDocumentId in originalDocument.GetLinkedDocumentIds())
                 {
                     var linkedDocument = solution.GetRequiredDocument(linkedDocumentId);
+
+                    // It's possible for us to have a solution snapshot where only part of a linked set of documents has
+                    // been updated.  As such, the other linked docs may have different contents/sizes than the original
+                    // doc we started with.  Skip those files as there's no sensible way to say that we have linked
+                    // symbols here when the contents are not the same.
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.cs:279]: ❔ Can we take this one step further and check that the `SourceText` is the same? While the syntax trees might differ, I would expect linked documents to at least share the `SourceText` internally.

File: src/EditorFeatures/Test2/FindReferences/FindReferencesTests.vb [hunk 2]
```diff
@@ -513,5 +514,69 @@ Namespace Microsoft.CodeAnalysis.Editor.UnitTests.FindReferences
...
+                Dim semanticModel1 = Await document1.GetSemanticModelAsync()
+                Dim root1 = Await semanticModel1.SyntaxTree.GetRootAsync()
+                Dim declarator1 = root1.DescendantNodes().OfType(Of VariableDeclaratorSyntax).First()
+                Dim symbol1 = semanticModel1.GetDeclaredSymbol(declarator1)
+                Assert.NotNull(symbol1)
+
+                Dim linkedSymbols = Await SymbolFinder.FindLinkedSymbolsAsync(symbol1, solution, cancellationToken:=Nothing)
+                Assert.Equal(2, linkedSymbols.Length)
+
+                ' now change the linked file and run again.
```
Comment (sharwell) [inline src/EditorFeatures/Test2/FindReferences/FindReferencesTests.vb:561]: 💡 Can you make a change where the file length is the same, but the identifier itself changed?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test2/FindReferences/FindReferencesTests.vb:561]: sure.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.cs:279]: i'd prefer to not accidentally cause us to break where someone legally updates the different forks with different source-texts :)

Comment (sharwell) [reply src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.cs:279]: Is there another way to do content validation efficiently? Or perhaps the number of cases of mismatched SourceText is so rare that we can do checksum comparison if the length is the same but the instance is different?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.cs:279]: generally speaking, i think this check, along with the check below (For name/kind) is sufficient.

File: src/EditorFeatures/Test2/FindReferences/FindReferencesTests.vb (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test2/FindReferences/FindReferencesTests.vb:556]: done.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.cs [hunk 3]
```diff
@@ -271,7 +272,15 @@ public static IEnumerable<TSymbol> FindSimilarSymbols<TSymbol>(TSymbol symbol, C
                 foreach (var linkedDocumentId in originalDocument.GetLinkedDocumentIds())
                 {
                     var linkedDocument = solution.GetRequiredDocument(linkedDocumentId);
+
+                    // It's possible for us to have a solution snapshot where only part of a linked set of documents has
```
Comment (genlu) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.cs:276]: >b [](http://example.com/codeflow?start=33&length=1)

I'm wondering if we need to handle this in completion as well. Here for example: [https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Completion/Providers/AbstractSymbolCompletionProvider.cs,324](https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Completion/Providers/AbstractSymbolCompletionProvider.cs,324)

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.cs:276]: ANy place that assume the files must have the same content is definitely in error and needs to be fixed up.

**Review by genlu — APPROVED** [review]
:shipit:

**Pull request merged**